### PR TITLE
feat(backend): OpenAPI migration foundation (U1 spike + U2 dashboard plumbing)

### PR DESCRIPTION
**Pull request by unclesp1d3r**
## Summary

Foundation work for the OpenAPI auto-generation migration. Lands **U1** (`@hono/zod-openapi@1.4.0` compatibility spike) and **U2** (dashboard surface plumbing) from `docs/plans/2026-05-31-001-feat-openapi-auto-generation-migration-plan.md`. The dashboard now has a runtime-emitted OpenAPI 3.1 spec endpoint at `/api/v1/dashboard/openapi.json` (anonymously fetchable, **eagerly cached at boot in production**), backed by shared response components and a `SessionCookie` security scheme that mirror the hand-rolled `dashboard-api.yaml`.

**Impact:** 24 files changed (+1,251 / -54). 8 new files, 16 modified (13 are pure 1-line router-class renames).
**Risk:** 🟢 Low — additive infrastructure, no behavioral change on any existing route, `OpenAPIHono extends Hono` preserves all `.get/.post/.use` semantics.
**Review time:** ~25 minutes — most of the diff is mechanical (13 router renames + 4 new modules with heavy test coverage).

## Commit map

| Commit | Scope | What landed |
|---|---|---|
| `d91a988` | feat(backend) | **U1** — `@hono/zod-openapi@^1.4.0` dependency + 7-case compatibility spike test |
| `b25920e` | feat(backend) | **U2** — 3 OpenAPI helper modules + dashboard aggregator + 13 router renames + `index.ts` consolidation |
| `fde099d` | fix(openapi) | `BetterAuthSession` → `SessionCookie` scheme name alignment to match the YAML |
| `2da14bc` | fix(review) | Multi-agent review autofix: smart-quote, eager spec generation, duplicate-registration guards, typed `sharedResponse()` helper, single-source response refs, comment sweep, 11-case helper test |
| `1294d21` | fix(test) | CodeRabbit review feedback: isolated prod-mode cache test using `*_TEST_ISOLATED=1` pattern |

### U1 — Compatibility spike (`d91a988`)

Adds `@hono/zod-openapi@^1.4.0` and a 7-case spike test (`packages/backend/tests/unit/openapi-spike.test.ts`) that validates the library invariants every downstream unit will rely on, before any route file is touched:

- OpenAPI 3.1 document round-trip via `getOpenAPI31Document`.
- `oneOf` emission for Zod discriminated unions with `unionPreferredType: 'oneOf'` — required for parity with the agent surface's discriminated unions (task report variants, agent error envelope).
- `.openapi('Name')` component registration producing clean `$ref`s.
- Shared response component `$ref` escape hatch.
- Per-surface security scheme round-trip through `createRoute.security` for all three schemes (cookie, API key, bearer).
- `z.literal` narrowing preservation in request bodies.
- Sub-app registry merging when a parent `OpenAPIHono` mounts a child — the load-bearing assumption for the dashboard aggregator.

Retained as a permanent regression smoke. Removal trigger documented in the file header.

#### New helpers under `packages/backend/src/openapi/`

- **`security.ts`** — `registerDashboardSecurity` / `registerControlSecurity` / `registerAgentSecurity`. Each throws on duplicate registration so two surfaces sharing a registry fails loudly at boot rather than silently last-writer-wins.
- **`components.ts`** — `registerDashboardResponseComponents` registers the four shared dashboard error responses (`AuthRequired`, `Forbidden`, `ValidationFailed`, `ResourceNotFound`). Exports `DASHBOARD_RESPONSE_REFS` (derived from a single-source `as const` array — typo at either site is a compile error) and `sharedResponse(ref)` (typed escape hatch so the unsafe `$ref` cast lives in exactly one place).
- **`spec-cache.ts`** — `mountCachedSpec` eagerly generates the spec at mount time in production (boot-time failure of a malformed Zod schema, logged with `OPENAPI_SPEC_GENERATION_FAILED` errorId, served as a typed dashboard envelope on failure). Dev/test regenerate per request for hot-reload.

#### New aggregator: `packages/backend/src/routes/dashboard/index.ts`

Owns the `OpenAPIHono` instance, registers security + components, mounts the cached `/openapi.json` endpoint **before** any auth middleware (so codegen tooling and Swagger UI can fetch it anonymously), and composes the twelve domain routers.

#### Mechanical changes

All 13 dashboard router files switch from `new Hono<AppEnv>()` to `new OpenAPIHono<AppEnv>()`. `OpenAPIHono extends Hono`, so every existing `.get/.post/.use` call continues to work unchanged. The conversion to `.openapi(createRoute(...), handler)` per route is U3's work, deferred to a follow-on PR.

`packages/backend/src/index.ts` reduces by 30 lines: twelve sibling `app.route('/api/v1/dashboard/<x>', ...)` calls collapse into one `app.route('/api/v1/dashboard', dashboardSurface)`. Per-surface CSRF middleware stays at the parent app level — paths still resolve correctly.

### Review-pass fixes (`fde099d`, `2da14bc`, `1294d21`)

Three review-fix commits address findings from a multi-agent review pass (`code-reviewer`, `pr-test-analyzer`, `comment-analyzer`, `type-design-analyzer`, `silent-failure-hunter`) plus follow-on CodeRabbit feedback:

- Security scheme name aligned to YAML (`SessionCookie` / `hh.session_token`).
- Smart-quote `'` (U+2019) replaced with ASCII per the no-fancy-punctuation rule.
- `mountCachedSpec` now eager-generates in production with try/catch + structured logging + typed error envelope on failure — a malformed Zod schema fails at boot, not at first client poll.
- Duplicate component / security-scheme registration guards throw loudly so a future caller wiring two surfaces onto a shared registry can't silently last-writer-wins.
- `DASHBOARD_RESPONSE_REFS` derived from a single-source array; typo at either site is a compile error.
- Typed `sharedResponse(ref): ResponseConfig` helper centralizes the documented `$ref` escape hatch so the `as any` cast lives in exactly one place.
- Comments swept to lead with invariants instead of citing migration-plan decision IDs (the plan file is gitignored per repo convention) or YAML line numbers (rot-prone once the YAML is deleted in a future PR).
- **New test file `packages/backend/tests/unit/openapi-helpers.test.ts`** (243 LOC, 11 cases) — covers the U2 plumbing-test gap: spec-endpoint anonymity, all four shared responses register, security-scheme shapes, duplicate-registration guards, `sharedResponse()` round-trip, cache identity.
- **New isolated test `packages/backend/tests/unit/openapi-spec-cache-prod.test.ts`** (112 LOC, 3 cases) — uses the repo's `*_TEST_ISOLATED=1` + custom preload pattern so `env.NODE_ENV='production'` snapshots correctly. Asserts eager generation at mount (no first-hit latency), identical bytes across requests, and the generator-call-count invariant (called exactly once, no per-request regeneration).

## What's NOT in this PR — remaining migration work

By design, ships U1 + U2 only. The plan defines 7 implementation units; U3-U7 are a follow-up effort:

- **U3** — Convert ~50 dashboard routes from `.get/.post` to `.openapi(createRoute(...), handler)` with `@hashhive/shared` Zod request/response schemas attached. Blocked on resolving the `.openapi()` extension propagation from schemas constructed with vanilla `z` in `@hashhive/shared` — likely via `OpenAPIRegistry.register(name, schema)` in each surface aggregator.
- **U4** — Write `packages/backend/scripts/diff-openapi-spec.ts`, iterate U3 until zero structural diff vs `dashboard-api.yaml`, then delete the YAML + the 627-line `dashboard-api-contract.test.ts` + the diff script in one commit.
- **U5** — Same pattern as U3 for the 11 control route files (~27 paths). RFC 9457 problem-details response shape. Register `ControlApiKey` scheme.
- **U6** — Migrate `packages/backend/src/routes/agent/index.ts` carefully (the "must never break" surface). Handle the discriminated unions via `unionPreferredType: 'oneOf'`. Preserve the agent error envelope byte-for-byte.
- **U7** — Strike the OpenAPI-sync rule from AGENTS.md; update README, ARCHITECTURE, CONTRIBUTING; collapse the dashboard read-endpoint contract from 4 pillars to 3; mark `shared-zod-openapi-wire-contract-mirror-2026-05-25.md` as superseded; delete `packages/openapi/` directory; cleanup `.coderabbit.yaml`.

The plan file lives at `docs/plans/2026-05-31-001-feat-openapi-auto-generation-migration-plan.md` (gitignored per repo convention). Future `/lfg` invocations on the same plan pick up U3+.

### 🆕 New (8 files)

```
packages/backend/src/openapi/
├── security.ts          ← per-surface scheme registration with duplicate guards
├── components.ts        ← shared response components, sharedResponse() helper
└── spec-cache.ts        ← eager prod cache + typed-envelope failure mode

packages/backend/src/routes/dashboard/
└── index.ts             ← aggregator owning the OpenAPIHono surface

packages/backend/tests/
├── preload-prod.ts                       ← NODE_ENV=production preload
├── unit/openapi-spike.test.ts            ← 7 library-invariant assertions
├── unit/openapi-helpers.test.ts          ← 11 plumbing assertions
└── unit/openapi-spec-cache-prod.test.ts  ← 3 isolated prod-cache assertions
```

### ✏️ Modified (16 files)

| File | Change |
|---|---|
| `packages/backend/src/index.ts` | -30 LOC: 12 sibling mounts collapsed into 1 aggregator |
| `packages/backend/src/routes/dashboard/*.ts` (13 files) | 1-line each: `new Hono<AppEnv>()` → `new OpenAPIHono<AppEnv>()` |
| `packages/backend/package.json` | +1 dep (`@hono/zod-openapi`), +1 isolated test entry |
| `bun.lock` | Lockfile resolution |

## Test plan

- [x] `just check` green (format + lint + type-check + build) on every commit
- [x] `just ci-check` green (full test suite) on every commit
- [x] U1 spike: 7/7 library-invariant assertions pass against `@hono/zod-openapi@1.4.0`
- [x] U2 helpers: 11/11 plumbing assertions pass (security scheme shape, shared-response registration, duplicate guards, `sharedResponse()` round-trip, spec-mount anonymity, cache identity)
- [x] U2 prod cache (isolated): 3/3 assertions pass under `NODE_ENV=production` preload — eager generation at mount, identical bytes across requests, generator called exactly once
- [x] No existing dashboard test required modification (`OpenAPIHono extends Hono` preserves the `.get/.post/.use` surface)
- [x] Security scheme name `SessionCookie` + cookie name `hh.session_token` match `packages/openapi/dashboard-api.yaml`
- [x] CodeRabbit review threads resolved (1/1) and review-body nitpicks replied (1/1)

## Risk assessment

| Factor | Score | Notes |
|---|---|---|
| **Size** | 🟢 Low | 24 files; 13 are pure renames; only 4 new modules carry novel logic |
| **Behavioral risk** | 🟢 Low | `OpenAPIHono extends Hono` — every existing route still resolves through the same `.get/.post/.use` middleware chain. Zero behavioral assertions changed. |
| **Test coverage** | 🟢 Strong | 21 new test cases across 3 files. Includes isolated prod-mode coverage that actually exercises the eager-cache path |
| **Dependencies** | 🟢 Low | 1 new dep (`@hono/zod-openapi@^1.4.0`); compatibility validated by the U1 spike |
| **Security** | 🟢 Low | `SessionCookie` scheme mirrors the existing dashboard contract; spec endpoint anonymously fetchable by design (metadata, not data) |
| **Reversibility** | 🟢 Easy | Revert lands cleanly — no DB migrations, no contract changes, no consumer-facing breaks |

**Mitigations in place:**
- Duplicate-registration guards on every component / scheme registrar throw loudly at boot.
- Spec generation failure is logged with stable errorId (`OPENAPI_SPEC_GENERATION_FAILED`) and returns a typed dashboard envelope rather than Hono's generic 500.
- Isolated prod-cache test pins eager-generation behavior; a future change that breaks the cache invariant fails CI.

## Reviewer checklist

**Setup & structural**
- [ ] `@hono/zod-openapi@^1.4.0` pinned in `packages/backend/package.json`; peer-deps satisfied (`hono >= 4.10.0` ✓, `zod ^4.0.0` ✓)
- [ ] All 13 dashboard router files use `new OpenAPIHono<AppEnv>()` (1-line change; `OpenAPIHono extends Hono` so `.get/.post/.use` semantics are preserved)
- [ ] `packages/backend/src/index.ts` mounts the dashboard aggregator at `/api/v1/dashboard` — 12 sibling mounts correctly collapsed; CSRF `requireSameOrigin()` filters still cover each domain path

**Plumbing modules**
- [ ] `security.ts` registers `SessionCookie` with cookie name `hh.session_token` matching `packages/openapi/dashboard-api.yaml#L1704-1708`
- [ ] `components.ts`: `DASHBOARD_RESPONSE_REFS` and the registration loop derive from a single `DASHBOARD_RESPONSE_NAMES` array
- [ ] `sharedResponse(ref): ResponseConfig` is the only place the `$ref` cast lives
- [ ] `spec-cache.ts` eager-generates in production, wraps generation in try/catch with the `OPENAPI_SPEC_GENERATION_FAILED` errorId
- [ ] Duplicate-registration guards throw loudly on both `responses` and `securitySchemes` re-registration

**Aggregator**
- [ ] Spec endpoint mounts BEFORE `surface.route(...)` so per-router `requireSession` middleware does NOT gate `/openapi.json`
- [ ] Root `app` in `index.ts` stays plain `Hono` (so the agent and control specs remain isolated when U5/U6 land)

**Tests**
- [ ] `openapi-spike.test.ts` retention rationale is documented at the file header
- [ ] `openapi-helpers.test.ts` covers spec anonymity, shared responses, scheme shapes, duplicate guards
- [ ] `openapi-spec-cache-prod.test.ts` is isolation-guarded (`describe.skipIf(!isIsolated)`) and wired into `package.json`'s test chain
- [ ] `preload-prod.ts` mirrors `preload.ts` minus `NODE_ENV='production'`

**Project guidelines (AGENTS.md)**
- [ ] Wire shapes still live in `@hashhive/shared` (no new local interface declarations for cross-boundary shapes)
- [ ] No fancy punctuation (smart quotes, em-dashes) in new files
- [ ] All files under the 800-line cap

## Follow-up coordination

U3 will start by resolving the `.openapi()` propagation question: the cleanest path is using `surface.openAPIRegistry.register(name, schema)` per aggregator for `@hashhive/shared` schemas (avoids coupling shared to `@hono/zod-openapi`, which would pull the dep into the frontend). The U1 spike already validates that registry-registered schemas round-trip through `$ref` correctly, so the migration pattern is unblocked.

---

[Compound Engineering](https://github.com/EveryInc/compound-engineering-plugin)
Claude Code

### CODE REVIEW

File: packages/backend/tests/unit/openapi-helpers.test.ts [hunk 1]
```diff
@@ -0,0 +1,243 @@
...
+      const body1 = await res1.text()
+      const body2 = await res2.text()
+      expect(body1).toBe(body2)
+    } finally {
+      if (originalEnv === undefined) {
+        delete process.env['NODE_ENV']
+      } else {
+        process.env['NODE_ENV'] = originalEnv
+      }
+    }
```
[the remark this answers is not in the document]
Comment (unclesp1d3r) [reply packages/backend/tests/unit/openapi-helpers.test.ts]: > production-mode caching test isn't exercising the prod branch — `mountCachedSpec` uses `env.NODE_ENV` from `packages/backend/src/config/env.ts`, which is computed once at module load.

Addressed in commit 1294d21. Adopted the repo's `*_TEST_ISOLATED=1` pattern (mirrors `tasks.test.ts`, `queue-manager.test.ts`):

- New `tests/preload-prod.ts` sets `NODE_ENV=production` before any project import so `env.ts` snapshots prod.
- New `tests/unit/openapi-spec-cache-prod.test.ts` runs three assertions the dev-mode test could not: (1) sanity-check `env.NODE_ENV` is actually production, (2) generator called eagerly at mount time (no first-hit latency), (3) generator called exactly once across N follow-up requests (real cache hit, not deterministic regeneration). Guarded with `describe.skipIf(!isIsolated)` so the catch-all `bun test` invocation silently no-ops.
- `package.json` test script wires `OPENAPI_SPEC_CACHE_PROD_TEST_ISOLATED=1 bun test --preload ./tests/preload-prod.ts ...` into the chain.
- Misleading prod test in `openapi-helpers.test.ts` dropped with a pointer comment so future readers find the isolated coverage.
